src/txmempool.cpp

```diff
@@ -618,6 +618,7 @@ void CTxMemPool::check(const CCoinsViewCache *pcoins) const
 
     if (GetRand(m_check_ratio) >= 1) return;
 
+    AssertLockHeld(::cs_main);
```

Comment:
```suggestion
    AssertLockHeld(::cs_main); // for GetSpendHeight
```

nit: Could mention for which function this is needed?

nit: Since all callers of `GetSpendHeight` already have cs_main, would it make sense to remove the recursive lock from `GetSpendHeight` itself and replace it with a debug-only/compile-only AssertLockHeld/EXCLUSIVE_LOCKS_REQUIRED?

Reply:
> Could mention for which function this is needed?

I've dug into this function a bit more, and I think it's more than that. `check()` takes a copy of the CoinsTip:

https://github.com/bitcoin/bitcoin/blob/3734adba390cef881445c4de780e2a3bb080c512/src/txmempool.cpp#L627

and is then fetching coins from that `CCoinsViewCache` via `CheckTxInputs()`:

https://github.com/bitcoin/bitcoin/blob/3734adba390cef881445c4de780e2a3bb080c512/src/txmempool.cpp#L610

If the coins in that `CCoinsView` were to be updated by a different thread while `check()` is running, then we could hit an assert. There's an implicit assumption that the `CCoinsView` won't change while this function is running, so I think we need `cs_main` throughout this function.

It's a shame that mempool calls back into validation in this way and has a circular dependency, but that's how it is for now.